src/Data/Id.hs

```diff
@@ -4,69 +4,172 @@
 import           Data.Binary (Binary)
 import           Data.Coerce (coerce)
 import           Data.Data
 import           Data.Hashable (Hashable)
 import           Data.OpenApi
+import           Data.String
+import           Data.Text as T
 import           Data.UUID (UUID)
 import qualified Data.UUID as UUID
 #ifndef ghcjs_HOST_OS
-import           Database.PostgreSQL.Simple.FromField (FromField)
-import           Database.PostgreSQL.Simple.ToField (ToField)
+import           Database.PostgreSQL.Simple.FromField as PG (FromField)
+import           Database.PostgreSQL.Simple.ToField as PG (ToField)
+import           Data.Csv as Csv hiding(Name)
 #endif
-import           GHC.Generics (Generic)
+import           Flat as F
+import           GHC.TypeLits
 import           Test.QuickCheck
 import           Web.HttpApiData
 import           Web.PathPieces (PathPiece(..))
 
 newtype Id t = Id { unId :: UUID }
-  deriving
-  ( Eq, Ord, Generic, Read, Show, Data
+  deriving stock Data
+  deriving newtype
+  ( Eq, Ord, Binary
 #ifndef ghcjs_HOST_OS
-  , ToField, FromField
+  , PG.ToField, PG.FromField
 #endif
-  , PathPiece, FromJSON
-  , ToJSON, NFData, Hashable, FromJSONKey, ToJSONKey, ToSchema, ToParamSchema
-  , FromHttpApiData, ToHttpApiData )
+  , FromJSON, ToJSON, NFData, Hashable, FromJSONKey, ToJSONKey, ToSchema
+  , ToParamSchema, FromHttpApiData, ToHttpApiData )
 
 type role Id nominal
 
+makePrisms ''Id
+
+mkId :: forall s. UUID -> Id s
+mkId = coerce
+{-# INLINE mkId #-}
+
+instance KnownSymbol s => Show (Id s) where
+  show (Id v) = "Id-" <> symbolVal (Proxy @s) <> "-" <> show v
+
+instance Flat (Id s) where
+  encode = F.encode . UUID.toWords . coerce
+  decode = coerce . (\(a,b,c,d) -> UUID.fromWords a b c d) <$> F.decode
+  size = F.size . UUID.toWords . coerce
+
 instance Arbitrary (Id t) where
   arbitrary = fmap Id $ UUID.fromWords
     <$> arbitrary
     <*> arbitrary
     <*> arbitrary
     <*> arbitrary
 
-instance Binary (Id t) where
+instance PathPiece (Id t) where
+  fromPathPiece = coerce . UUID.fromText
+  toPathPiece = UUID.toText . coerce
 
-instance PathPiece UUID.UUID where
-  fromPathPiece = UUID.fromText
-  toPathPiece = UUID.toText
+#ifndef ghcjs_HOST_OS
 
-makePrisms ''Id
+instance Csv.FromField (Id tag) where
+  parseField = parseField
+    >=> fmap mkId . maybe (fail "invalid UUID in CSV") pure . UUID.fromText
+
+instance Csv.ToField (Id tag) where
+  toField = Csv.toField . UUID.toText . coerce
+
+#endif
 
 -- | This is a more \"explicit\" 'coerce' specifically for 'Id'.
 -- You are forced to explicitly specify the phantom types you are converting
 -- via the @TypeApplications@ compiler extension.
-coerceId :: forall a b. Id (Ambiguous a) -> Id (Ambiguous b)
+coerceId, unsafeIdTagConvert :: forall a b. Id (Ambiguous a) -> Id (Ambiguous b)
 coerceId = coerce
 {-# INLINE coerceId #-}
 
+-- just another name for 'coerceId'
+unsafeIdTagConvert = coerce
```

Comment:
Make thing clear even without the comment. (if this typechecks)

```suggestion
unsafeIdTagConvert = coerceId
```

Reply:
This is not compiled. I think that using TA for this is too artificial. So I leave it as is (but I drop comments)